doc/boardState.mer

```diff
@@ -0,0 +1,10 @@
+---
+title: "Board states based on the player"
+---
+
+stateDiagram-v2
+    [*] --> Player?: Render()
```

Comment:
You're mixing state machines and flow charts together. The boxes should be possible states of some object, in this case the board, and the edges should be possible transitions between those states.

1. Decisions (e.g. `Player?`) are not states.
2. Actions (e.g. `UpdateBoard`) are not states.
3. I would expect the board to be able to directly or indirectly switch back and forth between the `AnswerBoard` and `PlainBoard` states, but instead there is no path from one to the other and in fact we always end up in the "trap state" of `UpdateBoard` with no transitions out.

I'd say your board object has two states, guesser view and code giver view. At the start of the game, the board is in the guesser view. There are multiple transitions that cause the board to switch from guesser view to code giver view and back.

If you need an example, refer to [this](https://github.com/EricSchrock/acquire/blob/main/doc/tile-state-machine.png) state machine from my Acquire game. Notice how the transitions show what is possible, given your current state. For example, once a tile is out of the bag, there is no way for it to get back in the bag. Once a tile is placed, it is placed for good and cannot transition to any other state. Tiles can switch back and forth between the `InHand` and `Selected` states.

Reply:
Okay I made some changes. A couple questions, are the parenthesis on the possible transitions mean those are real functions in your game? I didn't include the parenthesis in this updated chart since these are not real transition functions.

Also, I used `render()` as the start for my state diagram but technically render() is the function that gets called either time a board is made... Should I just get rid of that altogether?

Reply:
In the case of my example state machine, the transitions do happen to be methods on the [Tile](https://github.com/EricSchrock/acquire/blob/main/src/Tile.h) class. I think you've made the right choice to exclude parentheses in your case.

I do think you should remove `Render()` from the diagram. The dot clearly represents the starting point leading to the initial board state, so no need for text on that arrow.

Besides removing `Render()`, I'd say your board state machine looks great! Diagrams rarely capture the full complexity of the code so there are lots of choices you have to make on what to show and what not to show based on what you intend to communicate with the diagram. For example, below are two changes you could make to your diagram which may or may not be improvements, based on your goals. Try them out but only keep them if you like them.

1. You could add a transition called "Correct Guess" or "Guesses Left" or "Correct Guess and Guesses Left" that goes from the `GuesserView` state to the `GuesserView` state.
2. You could use four states instead of two (`TeamOneGuesserView`, `TeamOneCodeGiverView`, `TeamTwoGuesserView`, and `TeamTwoCodeGiverView`)

Reply:
Makes sense with Render().

I had thought about adding the point 1 you have to the chart but unfortunately the way Mermaid does it, it looks absolutely terrible sending the arrow like off to the side and skewing the whole graph...

I think with point 2, the two team's views aren't really different because the two codegivers should be able to look at the same board and both team's guessers will look at the same board at the same time as well so I think I'll leave it the way the diagram is now.